pubtools/_quay/push_docker.py

```diff
@@ -377,6 +380,89 @@ def rollback(self, backup_tags, rollback_tags):
+                List of operator push items.
+            existing_index_images: ([(digest, version)]):
+                List of tuple of digest + version(tag) of index images which existed
+                before new index image was pushed in the current task
+            backup_tags ({ImageData: str}):
+                Dictionary of ImageData (repo, tag, digest) -> manifest
+                holding containers which were overwritten in the currently running task
+            container_signature_handler (ContainerSignatureHandler):
+                ContanerSignatureHandler instance.
+        """
+        claim_messages = []
+        for item in push_items:
+            claim_messages += container_signature_handler.construct_item_claim_messages(item)
+        new_signatures = [(m["manifest_digest"], m["docker_reference"]) for m in claim_messages]
+        outdated_signatures = []
+
+        for image_data, manifest in backup_tags.items():
+            ext_repo = image_data.repo.split("/")[1].replace(INTERNAL_DELIMITER, "/")
+            if "manifests" in manifest:
+                for arch_manifest in manifest["manifests"]:
+                    outdated_signatures.append((arch_manifest["digest"], image_data.tag, ext_repo))
+            else:
+                outdated_signatures.append((image_data.digest, image_data.tag, ext_repo))
+
+        signatures_to_remove = []
+        for esig in container_signature_handler.get_signatures_from_pyxis(
+            [sig[0] for sig in outdated_signatures]
+        ):
+            if (
+                esig["manifest_digest"],
+                esig["reference"].split(":")[-1],
+                esig["repository"],
+            ) in outdated_signatures and (
+                esig["manifest_digest"],
+                esig["reference"],
+            ) not in new_signatures:
+                signatures_to_remove.append(esig["_id"])
+
+        container_signature_handler.remove_signatures_from_pyxis(signatures_to_remove)
+
+        for item in push_items:
+            claim_messages += container_signature_handler.construct_item_claim_messages(item)
+        new_signatures = [(m["manifest_digest"], m["docker_reference"]) for m in claim_messages]
+
+        signatures_to_remove = []
+        if existing_index_images:
+            outdated_signatures = []
+            operator_claim_messages = []
+            for item in operator_push_items:
+                operator_claim_messages += (
+                    container_signature_handler.construct_item_claim_messages(item)
+                )
+            new_operator_signatures = [
+                (m["manifest_digest"], m["docker_reference"]) for m in operator_claim_messages
+            ]
+            for esig in container_signature_handler.get_signatures_from_pyxis(
+                [digest_version[0] for digest_version in existing_index_images]
+            ):
+                if (esig["manifest_digest"], esig["reference"]) not in new_operator_signatures:
```

Comment:
Is it even possible for new signatures to have old manifest digest?

Reply:
In theory it is. In practice, hard to tell. Maybe opm also does modifications which includes timestamps